### PR TITLE
Retain only last copy of backup snapshot

### PR DESCRIPTION
**Pull request by cclhsu**
#### Proposed Changes ####

Retain only last copy of backup snapshot to resolve issue '[FEATURE] automatic successful backups should only retain the last snapshot'

#### Verification ####

Test case update in https://github.com/longhorn/longhorn-tests/pull/615

#### Linked Issues ####

Refs: https://github.com/longhorn/longhorn/issues/2231

#### Further Comments ####

None

Signed-off-by: Clark Hsu <clark.hsu@suse.com>

### CODE REVIEW

**Review by innobead — CHANGES_REQUESTED** [review]
Just changing here is misleading, and also not really meaningful (ex: always filtered by an empty filter). 

You should make sure the old snapshots purged except the latest one after the successful backup.

**Comment by cclhsu** [comment]
> Just changing here is misleading, and also not really meaningful (ex: always filtered by an empty filter).
> 
> You should make sure the old snapshots purged except the latest one after the successful backup.

Current Implementation will first get snapshots belong to recurring job and filter out the current successful snapshot for clean up.   Will need to separate the `snapshotAndCleanup` function so we do `snapshot -> backup -> cleanup backup -> cleanup snapshot`.

**Comment by innobead** [comment]
Thanks, @cclhsu explanation, LGTM. 

Leaving to @PhanLe1010 to approve because he is the original code change owner then we can merge.

**Review by innobead — APPROVED** [review]

**Review by shuo-wu — COMMENTED** [review]
Snapshot creation and cleanup happen before backup creation. If the last snapshot is removed here, can Longhorn still do delta backup later?

**Review by PhanLe1010 — CHANGES_REQUESTED** [review]
Like @shuo-wu mentioned, we want to to delta backup so before finishing taking backup, we need to keep the snapshot of the last backup and the snapshot of the current running backup. 

I think we can keep everything the same and just add 1 more step to clean up the snapshots after finish taking the backup at [here](https://github.com/longhorn/longhorn-manager/blob/ec1fa07891bf1a408024f74342d115ea3de01bef/app/volume.go#L582). It means that the flow is: take a new snapshot -> clean up to keep 2 snapshot (snapshot of last backup and snapshot of the current backup) -> taking backup -> clean up snapshots to keep only 1 snapshot of the current backup. This design is resilient when in the corner case: we don't create tons of snapshots when fail to taking backup (e.g., the backup target setting is empty)

**Review by PhanLe1010 — APPROVED** [review]
LGTM

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM, just minor comments to resolve.

File: app/volume.go [hunk 1]
```diff
@@ -310,10 +310,16 @@ func (job *Job) doRecurringSnapshot() (err error) {
...
 	}
 
-	return job.snapshotAndCleanup()
+	err = job.doSnapshot()
+	if err != nil {
+		return err
+	}
+
+	backupDone := false
+	return job.doSnapshotCleanup(backupDone)
```
[the remark this answers is not in the document]
Comment (cclhsu) [reply app/volume.go]: changed

File: app/volume.go [hunk 2]
```diff
@@ -495,7 +519,12 @@ func (job *Job) doRecurringBackup() (err error) {
 		return nil
 	}
 
-	if err := job.snapshotAndCleanup(); err != nil {
+	if err := job.doSnapshot(); err != nil {
+		return err
+	}
+
+	backupDone := false
+	if err := job.doSnapshotCleanup(backupDone); err != nil {
```
[the remark this answers is not in the document]
Comment (cclhsu) [reply app/volume.go]: changed

File: app/volume.go [hunk 3]
```diff
@@ -584,6 +613,11 @@ func (job *Job) doRecurringBackup() (err error) {
 		}
 		job.logger.Debugf("Cleaned up backup %v for %v", backup, volumeName)
 	}
+
+	backupDone = true
+	if err := job.doSnapshotCleanup(backupDone); err != nil {
```
[the remark this answers is not in the document]
Comment (cclhsu) [reply app/volume.go]: changed

File: app/volume.go [hunk 4]
```diff
@@ -335,13 +341,29 @@ func (job *Job) snapshotAndCleanup() (err error) {
 		}
 	}()
 
-	collection, err := volumeAPI.ActionSnapshotList(volume)
+	_, err = volumeAPI.ActionSnapshotList(volume)
 	if err != nil {
 		return err
 	}
```
[the remark this answers is not in the document]
Comment (cclhsu) [reply app/volume.go]: changed

**Review by innobead — APPROVED** [review]

**Pull request merged**